File: Clock/ViewController.m

#import "ViewController.h"

@interface ViewController ()
@property(weak,nonatomic)UIView *secondHandView;
@property(weak,nonatomic)UIView *minuteHandView;
@property(weak,nonatomic)UIView *hourHandView;


@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    CALayer *dialLayer= [[CALayer alloc] init];
    dialLayer.bounds = CGRectMake(0, 0, 150, 150);
    
    dialLayer.position = self.view.center;
    
    dialLayer.contents = (__bridge id _Nullable)([UIImage imageNamed:@"clock"].CGImage);
    
    [self.view.layer addSublayer:dialLayer];
    
    //    设置秒针
    UIView *secondHandView = [[UIView alloc]init];
    secondHandView.backgroundColor = [UIColor redColor];
    secondHandView.bounds = CGRectMake(0, 0, 1, 60);
    secondHandView.center = self.view.center;
    
    
    //    修改锚点
    secondHandView.layer.anchorPoint = CGPointMake(0.5, 1);
    [self.view addSubview:secondHandView];
    self.secondHandView = secondHandView;
    
    
    //    设置分针
    UIView *minuteHandView = [[UIView alloc]init];
    minuteHandView.backgroundColor = [UIColor darkGrayColor];
    minuteHandView.bounds = CGRectMake(0, 0, 3, 60);
    minuteHandView.center = self.view.center;
    
    
    //    修改锚点
    minuteHandView.layer.anchorPoint = CGPointMake(0.5, 1);
    [self.view addSubview:minuteHandView];
    self.minuteHandView = minuteHandView;
    
    //    设置时针
    UIView *hourHandView = [[UIView alloc]init];
    hourHandView.backgroundColor = [UIColor darkGrayColor];
    hourHandView.bounds = CGRectMake(0, 0, 4, 50);
    hourHandView.center = self.view.center;
    
    
    //    修改锚点
    hourHandView.layer.anchorPoint = CGPointMake(0.5, 1);
    [self.view addSubview:hourHandView];
    self.hourHandView = hourHandView;
    
    
    //    创建CADisplayLink
    CADisplayLink *link = [CADisplayLink displayLinkWithTarget:self selector:@selector(clockRunning)];
    //    将创建的CADisplayLink加入到主线程中
    [link addToRunLoop:[NSRunLoop mainRunLoop] forMode:NSDefaultRunLoopMode];
}

- (void)clockRunning{
    
    NSTimeZone *tZone = [NSTimeZone localTimeZone];
    
    
    //    获取日历
    NSCalendar *calendar = [NSCalendar currentCalendar];
    
    //    calendar.timeZone = tZone;
    
    NSDate *currentDate = [NSDate date];
    
    [calendar setTimeZone:tZone];
    
    NSDateComponents *currentTime = [calendar components:NSCalendarUnitSecond|NSCalendarUnitMinute|NSCalendarUnitHour|NSCalendarUnitTimeZone fromDate:currentDate];
    
    
    NSDateFormatter *nsdf2=[[NSDateFormatter alloc] init];
    [nsdf2 setDateStyle:NSDateFormatterShortStyle];
    [nsdf2 setDateFormat:@"ssSSSS"];
    NSString *t2=[nsdf2 stringFromDate:[NSDate date]];
    long curr=[t2 longLongValue];
    
    CGFloat angle = (M_PI * 2 / 60) * curr/10000 ;
    
    self.secondHandView.transform = CGAffineTransformMakeRotation(angle);
    
    CGFloat minuteAngle = (M_PI * 2 / 60) * currentTime.minute;
    self.minuteHandView.transform = CGAffineTransformMakeRotation(minuteAngle);
    
    CGFloat hourAngle = (M_PI * 2 / 12) * currentTime.hour;
    self.hourHandView.transform = CGAffineTransformMakeRotation(hourAngle);
}

@end

File: Clock/ClockSwift/ViewController.swift
//
//  ViewController.swift
//  ClockSwift
//
//  Created by Stan on 2017-02-03.
//  Copyright © 2017 stan. All rights reserved.
//

import UIKit

class ViewController: UIViewController {
    
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let dialLayer = CALayer()
        dialLayer.bounds = CGRect(x: 0, y: 0, width: 150, height: 150)
        
        dialLayer.position = view.center
        dialLayer.contents = UIImage(named: "clock")?.cgImage
        
        
        view.layer.addSublayer(dialLayer)
        
        
        //设置秒针，并加到view中
        secondHandView.center = view.center
        view.addSubview(secondHandView)
        
        //设置分针，并加到view中
        minuteHandView.center = view.center
        view.addSubview(minuteHandView)
        
        //设置时针，并加到view中
        hourHandView.center = view.center
        view.addSubview(hourHandView)
        
        
        
        
        
        
        //    创建CADisplayLink
        let link = CADisplayLink(target: self, selector: #selector(clockRunning))
        //将创建的CADisplayLink加入到主线程中
        link.add(to: RunLoop.main, forMode: RunLoopMode.defaultRunLoopMode)
    }
    
    
    @objc private func clockRunning() {
        let tZone = NSTimeZone.local
        
        var calendar = NSCalendar.current
        
        let currentDate = Date()
        calendar.timeZone = tZone
        
        let currentTime = calendar.dateComponents([Calendar.Component.hour,Calendar.Component.minute,Calendar.Component.second], from: currentDate)
        
        //设置秒针
        let angle  = Double (currentTime.second!) * (Double.pi * 2.0 / 60)
        
        secondHandView.transform = CGAffineTransform(rotationAngle: CGFloat(angle))
        
        //设置分针
        let minuteAngle = Double (currentTime.minute!) * (Double.pi * 2.0 / 60)
        
        minuteHandView.transform = CGAffineTransform(rotationAngle: CGFloat(minuteAngle))
        
        //设置时针
        let hourAngle = Double (currentTime.hour!) * (Double.pi * 2.0 / 12)
        
        hourHandView.transform = CGAffineTransform(rotationAngle: CGFloat(hourAngle))
        
    }
    
    
    
    
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
    //    MARK: - 懒加载
    //秒针
    lazy var secondHandView : UIView = {
        //设置秒针
        let secondHandView = UIView()
        secondHandView.backgroundColor = UIColor.red
        secondHandView.bounds = CGRect(x: 0, y: 0, width: 1, height: 60)
        
        //修改锚点
        secondHandView.layer.anchorPoint = CGPoint(x: 0.5, y: 1)
        return secondHandView
    }()
    
    
    //分针
    lazy var minuteHandView : UIView = {
        //设置分针针
        let minuteHandView = UIView()
        minuteHandView.backgroundColor = UIColor.darkGray
        minuteHandView.bounds = CGRect(x: 0, y: 0, width: 3, height: 60)
        
        //修改锚点
        minuteHandView.layer.anchorPoint = CGPoint(x: 0.5, y: 1)
        
        return minuteHandView
    }()
    
    
    //时针
    lazy var hourHandView : UIView = {
        //设置时针
        let hourHandView = UIView()
        hourHandView.backgroundColor = UIColor.darkGray
        hourHandView.bounds = CGRect(x: 0, y: 0, width: 1, height: 60)
        
        //修改锚点
        hourHandView.layer.anchorPoint = CGPoint(x: 0.5, y: 1)
        
        return hourHandView
    }()
    
}

